src/platform_impl/windows/ime.rs

```diff
@@ -0,0 +1,157 @@
+    ffi::{c_void, OsString},
+    mem::zeroed,
+    os::windows::prelude::OsStringExt,
+    ptr::null_mut,
+};
+
+use windows_sys::Win32::{
+    Foundation::POINT,
+    Globalization::HIMC,
+    UI::{
+        Input::Ime::{
+            ImmAssociateContextEx, ImmGetCompositionStringW, ImmGetContext, ImmReleaseContext,
+            ImmSetCandidateWindow, ATTR_TARGET_CONVERTED, ATTR_TARGET_NOTCONVERTED, CANDIDATEFORM,
+            CFS_EXCLUDE, GCS_COMPATTR, GCS_COMPSTR, GCS_CURSORPOS, GCS_RESULTSTR, IACE_CHILDREN,
+            IACE_DEFAULT,
+        },
+        WindowsAndMessaging::{GetSystemMetrics, SM_IMMENABLED},
+    },
+};
+
+use crate::{dpi::Position, platform::windows::HWND};
+
+pub struct ImeContext {
+    hwnd: HWND,
+    himc: HIMC,
+}
+
+impl ImeContext {
+    pub unsafe fn current(hwnd: HWND) -> Self {
+        let himc = ImmGetContext(hwnd);
+        ImeContext { hwnd, himc }
+    }
+
+    pub unsafe fn get_composing_text_and_cursor(
+        &self,
+    ) -> Option<(String, Option<usize>, Option<usize>)> {
+        if let Some(text) = self.get_composition_string(GCS_COMPSTR) {
+            if let Some(attrs) = self.get_composition_data(GCS_COMPATTR) {
+                let mut first: Option<usize> = None;
+                let mut last: Option<usize> = None;
+                let mut boundary_before_char = 0;
+
+                for (attr, chr) in attrs.into_iter().zip(text.chars()) {
+                    let char_is_targetted = attr as u32 == ATTR_TARGET_CONVERTED
+                        || attr as u32 == ATTR_TARGET_NOTCONVERTED;
+
+                    if first == None && char_is_targetted {
+                        first = Some(boundary_before_char);
+                    } else if first != None && last == None && !char_is_targetted {
+                        last = Some(boundary_before_char);
+                    }
+
+                    boundary_before_char += chr.len_utf8();
+                }
+                if first != None && last == None {
+                    last = Some(text.len());
+                } else if first == None {
+                    // IME haven't split words and select any clause yet, so trying to retrieve normal cursor.
+                    let cursor = self.get_composition_cursor(&text);
+                    first = cursor;
+                    last = cursor;
+                }
+
+                Some((text, first, last))
+            } else {
+                let cursor = self.get_composition_cursor(&text);
+                Some((text, cursor, cursor))
+            }
+        } else {
+            None
+        }
+    }
+
+    pub unsafe fn get_composed_text(&self) -> Option<String> {
+        self.get_composition_string(GCS_RESULTSTR)
+    }
+
+    unsafe fn get_composition_cursor(&self, text: &str) -> Option<usize> {
+        let cursor = ImmGetCompositionStringW(self.himc, GCS_CURSORPOS, null_mut(), 0);
+        if cursor >= 0 {
+            Some(text.chars().take(cursor as _).map(|c| c.len_utf8()).sum())
```

Comment:
I think you can make it into one line with `then`.

Reply:
To which variable can I use `then`? (the `cursor` is `i32`)

Reply:
`(cursor >= 0).then`

Reply:
done